tempesta_fw/cfg.h

```diff
@@ -0,0 +1,285 @@
+ *
+ * Each module defines a TfwCfgMod structure and calls tfw_cfg_mod_register().
+ * We maintain a list of registered modules in runtime and push events and
+ * configuration data via callbacks.
+ *
+ * @name is a unique text identifier of the module.
+ * Just like C identifiers, it must consist of alphanumeric characters and start
+ * with a letter and so on.
+ *
+ * @specs is the specification for the configuration parser that lists all
+ * possible configuration sections and directives for this module and describes
+ * how to handle them.
+ * @specs must be an array of TfwCfgSpec structures which is terminated
+ * by a null (zero'ed) element.
+ *
+ *
+ * @init and @exit callbacks act like module_init()/module_exit().
+ * They are called when a module is registered/unregistered and their only
+ * purpose is to eliminate some boilerplate function declarations and calls.
+ *
+ * @setup, @start, @stop and @cleanup callbacks are invoked when events
+ * are recevied via sysctl. The workflow is the following:
+ *   1. Receive a "start" event via syctl.
+ *   2. Invoke @setup.
+ *   3. Parse configuration, process @specs.
+ *   4. Invoke @start.
+ *   5. Receive a "stop" event via sysctl.
+ *   6. Invoke @stop.
+ *   7. Invoke @cleanup.
+ * Separate @setup/@start are needed to be able to run actions both before
+ * and after configuration parsing.
+ * Separate @stop/@cleanup callbacks are needed to avoid a scenario when a
+ * half of the modules are stopped and free()'d their memory, and another half
+ * is working and possibly references memory of the stopped modules. We stop
+ * work in @stop and free memory in @cleanup to avoid tricky synchronization.
+ *
+ * The only required field is the @name. All other fields may be NULL.
+ */
+typedef struct {
+	struct list_head list;	/* Private. Don't touch. */
+	const char *name;	/* [A-Za-z0-9_], starts with a letter. */
+	TfwCfgSpec *specs;	/* An array terminated by a null element. */
+	int  (*init)(void);
+	void (*exit)(void);
+	int  (*setup)(void);
+	int  (*start)(void);
+	void (*stop)(void);
+	void (*cleanup)(void);
+} TfwCfgMod;
+
+/* Subscribe/unsubscribe a module to sysctl events and configuration updates. */
+int tfw_cfg_mod_register(TfwCfgMod *mod);
+void tfw_cfg_mod_unregister(TfwCfgMod *mod);
+
+/*
+ * ------------------------------------------------------------------------
+ *	Generic TfwCfgSpec->handler functions.
+ * ------------------------------------------------------------------------
+ */
+
+typedef struct {
+	const char *name;
+	int value;
+} TfwCfgEnumMapping;
+
+int tfw_cfg_map_enum(const TfwCfgEnumMapping mappings[],
+		     const char *in_name, void *out_int);
+
+int tfw_cfg_check_single_val(const TfwCfgEntry *e);
+
+int tfw_cfg_parse_children(TfwCfgSpec *self, TfwCfgEntry *parsed_entry);
+
+int tfw_cfg_set_bool(TfwCfgSpec *self, TfwCfgEntry *parsed_entry);
+
+int tfw_cfg_set_int(TfwCfgSpec *spec, TfwCfgEntry *parsed_entry);
+
+typedef struct {
+	int is_multiple_of;
```

Comment:
Usially "is_" prefix means boolean type. However, you use it as integer (e.g. cache.c).